queries/locals.scm

```diff
@@ -10,4 +12,16 @@
 (record_pattern_argument pattern: (identifier) @local.definition)
 
 ; References
-(identifier) @local.reference
+(field_access
+  record: (identifier) @local.reference
+  field: (identifier) @local.reference)
+
+(function_call
+  function: (identifier) @local.reference)
+
+(argument
+  value: (identifier) @local.reference)
+
+(binary_expression (identifier) @local.reference)
+(case_subjects (identifier) @local.reference)
+(tuple (identifier) @local.reference)
```

Comment:
This is missing list elements, right? I'm a bit afraid that specifying each of these is a bit error-prone and may quickly become out-of-date as the language evolves. Are we able to specify where an `identifier` is _not_ going to be a reference (e.g. labels) instead?

I'm also open to changing the AST such that labels are actually a different node type, such as `(label)`. This adds a bit of semantic richness to the AST and it should help more than hurt syntax highlighting efforts (I think :sweat_smile:)

Reply:
Aha I knew I forgot one! :sweat_smile: 

I did try something like

```lisp
(field_access field: (identifier) @ignore)
(identifier) @local.reference
```

at first so we would only have to write a list of `(identifier)`s that we _don't_ want locals tracking on, but that doesn't appear to work.

On the one hand I do like a very generic `(identifier)` node in the tree, but I think you might be right: changing the record keys to a different node type would probably make things cleaner. I'll experiment! :scientist:

Reply:
> I do like a very generic `(identifier)` node in the tree

I must say that I do as well. After reviewing these changes again, I wondered if somehow we were taking a wrong turn. _That said_, I don't think we are. My thoughts on this are probably best embodied by an unrelated comment that I wrote a bit ago:
https://github.com/J3RN/tree-sitter-gleam/blob/b0a23fe88f5f6fa03d87a643e08a54b1aaadbbbd/grammar.js#L474-L475

The name "identifier" even makes perfect sense for "a variable or local function" as it _identifies_ them; perhaps the only more suitable name would be "name", since folks tend to say "function name" or "variable name" rather than "function identifier" or "variable identifier" (probably because those are a mouthful).

The only spots where we deviate from "an identifier is a variable or a function" (with constants, parameters, and assignments targets grouped in with "variables") (disregarding the record-access/remote-function ambiguity) is a couple of spots where we refer to module names as "identifier". However, I would argue that we should probably move to make those use the new `(module)` node type (probably by way of `alias($._name, $.module)` since the actual `$.module` definition is special-purpose).

**TL;DR**: I think we are moving from more ambiguity to less, and I think this is a great asset :100:

Reply:
If Gleam changes to allow statements such as this:
```gleam
import gleam/list

const mymod = list
```
Then I'll probably need to rearrange some things such that `(identifier)` means "function name _or_ variable name _or_ module name". We'll see :grin: